File: android/src/main/kotlin/cloud/trustpin/flutter/sdk/ErrorMapping.kt
package cloud.trustpin.flutter.sdk

import cloud.trustpin.kotlin.sdk.TrustPinError
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.launch

/** A unit of work produced by a handler's synchronous parse phase. */
internal typealias HandlerOperation = suspend () -> Any?

/**
 * Translates a [TrustPinError] case into the stable string code we send to
 * Dart. Unknown future cases fall through to `INVALID_PROJECT_CONFIG` —
 * matching the previous behaviour.
 */
internal fun mapTrustPinError(error: TrustPinError): String = when (error) {
    is TrustPinError.InvalidProjectConfig -> ErrorCode.INVALID_PROJECT_CONFIG
    is TrustPinError.AlreadyInitialized -> ErrorCode.ALREADY_INITIALIZED
    is TrustPinError.ErrorFetchingPinningInfo -> ErrorCode.ERROR_FETCHING_PINNING_INFO
    is TrustPinError.InvalidServerCert -> ErrorCode.INVALID_SERVER_CERT
    is TrustPinError.PinsMismatch -> ErrorCode.PINS_MISMATCH
    is TrustPinError.AllPinsExpired -> ErrorCode.ALL_PINS_EXPIRED
    is TrustPinError.ConfigurationValidationFailed ->
        ErrorCode.CONFIGURATION_VALIDATION_FAILED
    is TrustPinError.DomainNotRegistered -> ErrorCode.DOMAIN_NOT_REGISTERED
    is TrustPinError.Timeout -> ErrorCode.FETCH_CERTIFICATE_TIMEOUT
    is TrustPinError.ConfigIntegrityError -> ErrorCode.CONFIG_INTEGRITY_FAILED
    // Documented cross-platform contract: operations before setup surface as
    // INVALID_PROJECT_CONFIG, matching the iOS/macOS SDK which has no
    // separate not-initialized case.
    is TrustPinError.NotInitialized -> ErrorCode.INVALID_PROJECT_CONFIG
    is TrustPinError.SetupInProgress -> ErrorCode.SETUP_IN_PROGRESS
    is TrustPinError.LockTimeout -> ErrorCode.LOCK_TIMEOUT
    is TrustPinError.SSLContextSetupFailed -> ErrorCode.SSL_CONTEXT_SETUP_FAILED
    is TrustPinError.UnsupportedDevice -> ErrorCode.UNSUPPORTED_DEVICE
    else -> ErrorCode.INVALID_PROJECT_CONFIG
}

/**
 * Delivers any error thrown from a handler to the Flutter side using the
 * appropriate stable code. Returns `true` only for a plain
 * [CancellationException] (cooperative cancellation), so the caller can
 * re-throw it and honour structured concurrency. Returns `false` for every
 * other case, including [TimeoutCancellationException] — a timeout is a
 * terminal outcome of the handler, not a cancellation request directed at
 * the surrounding scope.
 *
 * Timeouts are enforced by the native SDK (surfacing as
 * [TrustPinError.Timeout]); the [TimeoutCancellationException] branch is
 * defensive, in case a `kotlinx.coroutines` timeout ever leaks through, and
 * must stay before its superclass [CancellationException].
 */
internal fun Result.deliverError(
    error: Throwable,
    defaultCode: String
): Boolean = when (error) {
    is TimeoutCancellationException -> {
        error(ErrorCode.FETCH_CERTIFICATE_TIMEOUT, "Timed out", null)
        false
    }
    is CancellationException -> {
        error(ErrorCode.CANCELLED, "Operation was cancelled", null)
        true
    }
    is InvalidArgumentsError -> {
        error(ErrorCode.INVALID_ARGUMENTS, error.message, null)
        false
    }
    is TrustPinError -> {
        error(mapTrustPinError(error), error.message, null)
        false
    }
    else -> {
        error(defaultCode, error.message, null)
        false
    }
}

/**
 * Runs a handler in two phases:
 *
 * 1. [parse] runs synchronously on the calling thread. It extracts arguments
 *    and returns the suspending closure to execute. Anything thrown here
 *    (e.g. [InvalidArgumentsError]) is delivered immediately.
 * 2. The returned [HandlerOperation] is launched on [scope]. Its result is
 *    delivered through [result]; recognised errors are mapped to stable codes.
 */
internal fun execute(
    scope: CoroutineScope,
    result: Result,
    defaultCode: String,
    parse: () -> HandlerOperation
) {
    val operation: HandlerOperation = try {
        parse()
    } catch (e: Exception) {
        result.deliverError(e, defaultCode)
        return
    }

    scope.launch {
        try {
            result.success(operation())
        } catch (e: Exception) {
            val shouldRethrow = result.deliverError(e, defaultCode)
            if (shouldRethrow) throw e
        }
    }
}

File: darwin/trustpin_sdk/Sources/trustpin_sdk/ErrorMapping.swift
#if canImport(Flutter)
@preconcurrency import Flutter
#elseif canImport(FlutterMacOS)
@preconcurrency import FlutterMacOS
#endif
import Foundation
import TrustPinKit

/// Translates a `TrustPinErrors` case into the stable string code we send to
/// Dart. Unknown future cases fall through to `INVALID_PROJECT_CONFIG` —
/// matching the previous behaviour.
func mapTrustPinError(_ error: TrustPinErrors) -> String {
    switch error {
    case .invalidProjectConfig:          return ErrorCode.invalidProjectConfig
    case .alreadyInitialized:            return ErrorCode.alreadyInitialized
    case .errorFetchingPinningInfo:      return ErrorCode.errorFetchingPinningInfo
    case .invalidServerCert:             return ErrorCode.invalidServerCert
    case .pinsMismatch:                  return ErrorCode.pinsMismatch
    case .allPinsExpired:                return ErrorCode.allPinsExpired
    case .configurationValidationFailed: return ErrorCode.configurationValidationFailed
    case .domainNotRegistered:           return ErrorCode.domainNotRegistered
    case .timeout:                       return ErrorCode.fetchCertificateTimeout
    case .configIntegrityFailed:         return ErrorCode.configIntegrityFailed
    @unknown default:                    return ErrorCode.invalidProjectConfig
    }
}

/// Converts any error thrown from a handler into a `FlutterError` with the
/// appropriate stable code and message.
///
/// Timeouts are enforced by the native SDK and arrive as
/// `TrustPinErrors.timeout`, mapped like every other SDK error case.
///
/// - parameters:
///   - error: the thrown error
///   - defaultCode: code used for generic, unclassified errors
func flutterError(
    from error: Error,
    defaultCode: String
) -> FlutterError {
    switch error {
    case is CancellationError:
        return FlutterError(
            code: ErrorCode.cancelled,
            message: "Operation was cancelled",
            details: nil
        )
    case let invalid as InvalidArgumentsError:
        return FlutterError(
            code: ErrorCode.invalidArguments,
            message: invalid.message,
            details: nil
        )
    case let trustPin as TrustPinErrors:
        return FlutterError(
            code: mapTrustPinError(trustPin),
            message: trustPin.localizedDescription,
            details: nil
        )
    default:
        return FlutterError(
            code: defaultCode,
            message: error.localizedDescription,
            details: nil
        )
    }
}

// MARK: - Execute helper

/// The async operation produced by a handler's synchronous "parse" phase.
/// Handlers return a certificate PEM string, a boolean (e.g.
/// `isConfigurationLoaded`), or no value.
typealias HandlerOperation = @Sendable () async throws -> Any?

/// Runs a handler in two phases:
///
/// 1. The caller's `parse` block runs synchronously. It extracts arguments
///    and returns the async closure to execute. Anything thrown here (e.g.
///    `InvalidArgumentsError`) is delivered immediately as a Flutter error.
/// 2. The returned async closure runs on an unstructured `Task`. Its result
///    is delivered through the supplied `ResultBox`. All recognised error
///    types are mapped to stable Flutter error codes.
func execute(
    _ box: ResultBox,
    defaultCode: String,
    _ parse: () throws -> HandlerOperation
) {
    let operation: HandlerOperation
    do {
        operation = try parse()
    } catch {
        box.deliver(flutterError(from: error, defaultCode: defaultCode))
        return
    }

    Task { [box] in
        do {
            try Task.checkCancellation()
            let value = try await operation()
            try Task.checkCancellation()
            box.deliver(value)
        } catch {
            box.deliver(flutterError(from: error, defaultCode: defaultCode))
        }
    }
}
